modules/rosapps/applications/devutils/shlextdbg/shlextdbg.cpp

```diff
@@ -7,27 +7,380 @@
+    wprintf(L"    dll: Optional dllname to create the object from, instead of CoCreateInstance\n");
+    wprintf(L"    filename: The filename to pass to IShellExtInit->Initialze\n");
+    wprintf(L"    shlextype: The type of shell extention to run:\n");
+    wprintf(L"               /IShellPropSheetExt to create a property sheet\n");
+    wprintf(L"               /IContextMenu=verb to activate the specified verb\n");
+    wprintf(L"    waitoptions: Specify how to wait:\n");
+    wprintf(L"                 /explorerinstance: Wait for SHGetInstanceExplorer (Default)\n");
+    wprintf(L"                 /infinite: Keep on waiting infinitely\n");
+    wprintf(L"                 /openwindows: Wait for all windows from the current application to close\n");
+    wprintf(L"                 /input: Wait for input\n");
+    wprintf(L"                 /nowait\n");
+    wprintf(L"\n");
+    wprintf(L"shlextdbg /shgfi=path\n");
+    wprintf(L"    Call SHGetFileInfo. Prefix path with $ to parse as a pidl.\n");
+    wprintf(L"\n");
+    wprintf(L"shlextdbg /assocq <[{bhid}]path> <string|data|key> <type> <initflags> <queryflags> <initstring> [extra] [maxsize]\n");
+    wprintf(L"    Uses the default implementation from AssocCreate if path is empty.\n");
+    wprintf(L"\n");
+    wprintf(L"shlextdbg /shellexec=path [/see] [verb] [class]\n");
+    wprintf(L"\n");
+    wprintf(L"shlextdbg /dumpmenu=[{clsid}]path [/cmf]\n");
+}
+
+/*
+Examples:
+
+/clsid={513D916F-2A8E-4F51-AEAB-0CBC76FB1AF8} /IShellExtInit=C:\RosBE\Uninstall.exe /IShellPropSheetExt
+/clsid=CompressedFolder /IShellExtInit=e:\test.zip /IContextMenu=extract /openwindows
+/clsid=CompressedFolder /IShellExtInit=e:\test.zip /IContextMenu=extract /openwindows /dll=R:\build\dev\devenv\dll\shellext\zipfldr\Debug\zipfldr.dll
+/shgfi=c:\freeldr.ini
+/assocq "" string 1 0 0 .txt
+/assocq "" string friendlytypename 0x400 0 .txt "" 10
+/openwindows /shellexec=c: /invoke properties
+/dumpmenu=%windir%\explorer.exe /extended
+/dumpmenu {D969A300-E7FF-11d0-A93B-00A0C90F2719}c:
+
+*/
+
+static LONG StrToNum(LPCWSTR in)
+{
+    WCHAR *end;
+    LONG v = wcstol(in, &end, 0);
+    return end > in ? v : 0;
+}
+
+static int ErrMsg(int Error)
+{
+    WCHAR buf[400];
+    for (UINT e = Error, cch;;)
+    {
+        lstrcpynW(buf, L"?", _countof(buf));
+        cch = FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS, 0, e, 0, buf, _countof(buf), NULL);
+        for (; cch && buf[cch - 1] <= ' ';)
+            buf[--cch] = UNICODE_NULL; // Remove trailing newlines
+        if (cch || HIWORD(e) != HIWORD(HRESULT_FROM_WIN32(1)))
+            break;
+        e = HRESULT_CODE(e); // "WIN32_FROM_HRESULT"
+    }
+    wprintf(Error < 0 ? L"Error 0x%.8X %s\n" : L"Error %d %s\n", Error, buf);
+    return Error;
+}
+
+template<class T> static bool CLSIDPrefix(T &String, CLSID &Clsid)
+{
+    WCHAR buf[39];
+    if (String[0] == '{')
+    {
+        lstrcpynW(buf, String, 39);
```

Comment:
I'm hesitating with this
```suggestion
        lstrcpynW(buf, String, _countof(buf));
```
but I suppose this "39" is equal to the length of string representation of a CLSID, including its opening and closing brackets?

Reply:
38 + null terminator, yes.